File: C#/RPC/Hayaa.RPC/Client/ServiceMethdoProxy.cs
using Hayaa.RPC.Common.Config;
using Hayaa.RPC.Service.Protocol;
using Hayaa.RPC.Service.Util;
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace Hayaa.RPC.Service.Client
{
    public class ServiceMethdoProxy
    {
        public static T Invoke<T>(String interfaceName, String methodName, Dictionary<String, RpcDataValue> paramater)
        {
            T result = default(T);
            try
            {
                String msgID = Guid.NewGuid().ToString("N");
               
                ResultMessage msgResult = null;
                int timeOut = ConfigHelper.Instance.GetComponentConfig().SessionTimeout;
                int time = 0;
                 ClientHelper.Instance.EnQueue(new Protocol.MethodMessage()
                {
                    InterfaceName = interfaceName,
                    Method = methodName,
                    Paramater = paramater,
                    MsgID = msgID
                });
                Console.WriteLine("result wait timeOut:" + timeOut);
                while (time<timeOut)
                {
                    Console.WriteLine("result wait loop");
                    msgResult = ClientHelper.Instance.GetResult(msgID);
                    if (msgResult != null)
                    {
                        time= timeOut+1;
                    }
                    if(time< timeOut)
                    Thread.SpinWait(1);
                    time= time+3;//考虑代码操作时间增量
                }
                if (msgResult != null)
                {
                    if (String.IsNullOrEmpty(msgResult.ErrMsg))
                        result = JsonHelper.Deserialize<T>(msgResult.Result);
                }
                else
                {
                    ClientHelper.Instance.DelTimeoutMsgID(msgID);
                }
                Console.WriteLine("Server result:" + msgResult.Result);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            return result;
        }

    }
}

File: Java/v1/src/main/java/hayaa/rpc/client/ServiceMethdoProxy.java
package hayaa.rpc.client;

import hayaa.common.JsonHelper;
import hayaa.common.StringUtil;
import hayaa.rpc.common.config.RPCConfigHelper;
import hayaa.rpc.common.protocol.MethodMessage;
import hayaa.rpc.common.protocol.ResultMessage;
import hayaa.rpc.common.protocol.RpcDataValue;

import java.util.*;

public class ServiceMethdoProxy {
    public static Object invoke(String interfaceName, String methodName, List<RpcDataValue> paramater, Class<?> resultType) {
        String strResult = null;
        String msgID = UUID.randomUUID().toString();
        MethodMessage methodMessage = new MethodMessage();
        methodMessage.setInterfaceName(interfaceName);
        methodMessage.setMethod(methodName);
        if(paramater.size()==0){
            paramater=null;
        }
        methodMessage.setParamater(paramater);
        methodMessage.setMsgID(msgID);
        ResultMessage msgResult = null;
        int timeOut = RPCConfigHelper.getRPCConfig().getSessionTimeout();
        int time = 0;
        try {

            Boolean action = ClientHelper.get_instance().enQueue(methodMessage);
            if (!action) {
                return null;
            }
            //System.out.println("result wait timeOut:" + timeOut);
            while (time < timeOut) {
                msgResult = ClientHelper.get_instance().GetResult(msgID);
                if (msgResult != null) {
                    time = timeOut + 1;
                }
                if (time < timeOut) {
                    Thread.sleep(1);
                }
                //考虑代码操作时间增量
                time = time + 3;
            }
            if (msgResult != null) {
                if (StringUtil.IsNullOrEmpty(msgResult.getErrMsg())) {
                    strResult = msgResult.getResult();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        //System.out.println("Server result:" + strResult);
        Object result = null;
        if (!StringUtil.IsNullOrEmpty(strResult)) {
            System.out.println("invoke done");
            result = JsonHelper.gsonDeserialize(strResult, resultType);
        }
        return result;
    }

}
